KEP-3178: Cleaning up IPTables Chain Ownership

danwinship opened the pull request:
- One-line PR description: Initial draft of KEP-3178, to clean up kubelet-vs-kube-proxy IPTables chain ownership

- Issue link: #3178

- Other comments: (I'm filing this now but targeting it at 1.25 for alpha given that it requires changes to the same parts of the iptables proxy that there are already multiple other pending KEP PRs for.)

aojea commented:
/cc

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,687 @@
+
+Is there any other related legacy networking cruft in kubelet that it
+would make sense to remove/move at the same time as the IPTables
+stuff?
```
```
$ git grep -i NonMasq
cmd/kubelet/app/options/options.go:     // nonMasqueradeCIDR configures masquerading: traffic to IPs outside this range will use IP masquerade.
cmd/kubelet/app/options/options.go:     NonMasqueradeCIDR string
cmd/kubelet/app/options/options.go:             NonMasqueradeCIDR:       "10.0.0.0/8",
cmd/kubelet/app/options/options.go:     fs.StringVar(&f.NonMasqueradeCIDR, "non-masquerade-cidr", f.NonMasqueradeCIDR, "Traffic to IPs outside this range will use IP masquerade. Set to '0.0.0.0/0' to never masquerade.")
cmd/kubelet/app/server.go:              s.NonMasqueradeCIDR)
pkg/kubelet/kubelet.go: nonMasqueradeCIDR string) error {
```

danwinship replied:
`git pull`
:slightly_smiling_face: 
(removed in #107096)

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,687 @@
+`KUBE-XLB-` chain for the service. At the end of the chain, it calls
+`KUBE-MARK-DROP` on any unmatched packets.
+
+To do this without `KUBE-MARK-DROP`, we simply remove the
```
I find the fact that the processing logic is in one place is convenient for comprehension, but I guess we can free up one bit in the mark...

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,687 @@
+port, and calling `-j DROP`. Any traffic that matched one of the
+allowed source ranges in the `KUBE-FW-` chain would have been DNAT'ed
+to a pod IP by this point, so any packet that still has the original
+load balancer IP as its destination when it reaches the `filter` table
```
Caution - some folks access service LB's from within the cluster (and want us to stop internally redirecting!!).  If you blindly drop, this will not work, I think.

danwinship replied:
OK, thinking about this some more, there really isn't much point in having the FW chain in `nat` anyway; we can just put all of the firewalling rules in the `filter` table and only the rewriting rules in the `nat` table.

(Note that firewalling currently happens before the ClusterIP redirect anyway, so it is already the case that if you want to connect to an LB IP from a pod, where the service uses LoadBalancerSourceRanges, then the source ranges need to include the pod IP. Our e2e tests require this behavior.)

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
````diff
@@ -0,0 +1,687 @@
+```
+<<[UNRESOLVED iptables-wrapper-help ]>>
+
+It might make sense to have kubelet continue to create a single
````
+1

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,687 @@
+post.)
+
+Kube-proxy will be updated to not use `KUBE-MARK-DROP`, as described
+above. (This change is unconditional; it is _not_ feature-gated.) We
```
why no gate?  What happens if we've done something disastrously dumb - how can a cluster back out of this change?

danwinship replied:
Since this just changes kube-proxy's internals without actually changing the visible functionality, I was considering it a cleanup/bugfix, not a feature. (So if we do something disastrously dumb, and somehow don't catch it, then affected users would have to downgrade, just like with any other major bug.)

If we feature-gate the kube-proxy change, then we'd have to push out the rest of the schedule by two releases, because you can't have the kubelet feature gate go to beta until 2 releases after kube-proxy starts not depending on `KUBE-MARK-DROP`. (So we'd have two separate feature gates, a lot like the "Move `KUBE-MARK-DROP` to Kube-Proxy Rather Than Removing It" example under "Alternatives".)

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,687 @@
+  1. It will stop creating `KUBE-MARK-DROP`, `KUBE-MARK-MASQ`,
+     `KUBE-FIREWALL`, and `KUBE-POSTROUTING`.
+
+  2. It will delete the `KUBE-MARK-DROP` and `KUBE-FIREWALL` chains,
```
Do we need to actively delete them?  Or can we just rely on reboots to phase them out?

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,687 @@
+
+As long as we wait 2 releases between Alpha and Beta, then it is
+guaranteed that when the user upgrades to Beta, the currently-running
+kube-proxy will be one that does not require the `KUBE-MARK-DROP`
```
What about people running other proxies?  E.g. cilium - it seems unlikely that literally cilium depends on iptables rules, but I've seen worse.  And there ARE other implementations out there..

danwinship replied:
Part of the work for the Alpha release is:

> We will document, in the appropriate places, the fact that external
components should not be assuming the existence of Kubernetes's
IPTables chains. (This probably includes a release note and a blog
post.)

(Though really this should be done well before the actual alpha release date, to give third-party developers time to get an updated release out in parallel with the `IPTablesCleanup` alpha.)

Also, the graduation criteria for Beta includes:

> - If the removal of the IPTables chain creation from kubelet has been
  found to cause problems with third-party components, then the SIG
  must agree that these problems can be addressed with release notes
  and do not justify delaying Beta. (eg, a bug involving a component
  which is not widely used, or which can be worked around by upgrading
  to a newer version of that component, or changing its
  configuration.)

(That's kind of badly-worded because this section is "criteria for proceeding", not "criteria for delaying", but the intent here is "if the SIG thinks the world isn't ready for this KEP to go to Beta, then it won't go to Beta".)

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,720 @@
+the `KUBE-SVC` or `KUBE-XLB` chain for the service. At the end
+of the chain, it calls `KUBE-MARK-DROP` on any unmatched packets.
+
+But there is no reason why the `FW` chain could not simply be moved to
```
I wonder if there's any side-effect here with third-party things that might add additional drop rules or change the default policy, which a unilateral ALLOW would break.

thockin replied:
Or maybe `-j RETURN` rather than allow?

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,720 @@
+ever seemed to care. Given the current state, in which only
+LoadBalancer traffic gets firewalled, the simplification from reusing
+the existing matching logic in the `filter` table seems to be
+outweighed by the complication of having to deal with
```
I'm 50/50 - having MARK-DROP in the nat chain puts all the logic in one place, which is nice.  That said, I buy the REJECT consistency below

thockin reviewed (commented):
LGTM except ACCEPT questions,

Thanks for this, long overdue

aojea commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,720 @@
+of the chain, it calls `KUBE-MARK-DROP` on any unmatched packets.
+
+But there is no reason why the `FW` chain could not simply be moved to
+the `filter` table (with calls to `-j ACCEPT` for allowed source
```
currently, KUBE-SERVICES is called from PREROUTING and OUTPUT ... these new filtering rules has to go to the filter tables on INPUT, FORWARDING and OUTPUT , right?

danwinship replied:
yes

danwinship replied:
meh, `nat` `PREROUTING` happens before anything in `filter`. One of these days I'll remember how iptables ordering works without having to google it...

@thockin didn't you have some kube-proxy iptables-processing-order infographic somewhere?

aojea replied:
I have this printed in my table 
![image](https://user-images.githubusercontent.com/6450081/155537288-bc8c8896-ef9f-469f-9069-47e3676b828f.png)

aojea replied:
:thinking: we miss the service IP on the filter tables, DNAT was already performed

danwinship replied:
Anyway, this worked right in the original proposal (keep the FW chain in NAT, fall through rather than KUBE-MARK-DROP at the end, have the KUBE-FIREWALL chain in filter drop any not-yet-DNAT-ed packets addressed to the LB IP). I changed it in part because of [Tim's comment here](https://github.com/kubernetes/enhancements/pull/3179#discussion_r808565783) but (a) as long as we keep doing the pod-to-LB-IP short-circuiting, the original proposal works, and (b) if we stop doing the short-circuiting in the future, we'd have to rewrite all the rules anyway, and we can just address this then.

aojea commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,720 @@
+  - `KUBE-MARK-DROP` and `KUBE-FIREWALL`
+
+    `KUBE-MARK-DROP` marks packets as needing to be dropped by setting
+    a _different_ configurable bit on the packet mark. `KUBE-FIREWALL`
```
KUBE-FIREWALL was also extended to deal with the route_localnet CVE
https://github.com/kubernetes/kubernetes/pull/91569

danwinship replied:
Yeah, forgot about that.

So this is its own mess: 91569 only talks about NodePort, but the problem theoretically applies to HostPorts too. Except, it only applies if you set `route_localnet`, which kube-proxy does and kubelet doesn't... so for our own purposes we can move 91569's rule from kubelet to kube-proxy. (In fact, it should have been there originally, but `KUBE-FIREWALL` sounds like a good place to put it, and that's created from kubelet, so it got put there.)

But given that it was previously in kubelet, there are probably alternate proxies and hostport implementations that would be vulnerable if we didn't have this rule... so we may have to keep creating it...

We should probably have a KEP about eventually getting rid of localhost NodePort/HostPort behavior...

aojea replied:
I have this printed in my table 
![image](https://user-images.githubusercontent.com/6450081/155537288-bc8c8896-ef9f-469f-9069-47e3676b828f.png)

aojea commented:
+1 but I suggest to postpone one release waiting for the dockershim feedback, it seems is causing more drama than expected and adding more changes on top can complicate it more

danwinship commented:
> +1 but I suggest to postpone one release waiting for the dockershim feedback, it seems is causing more drama than expected and adding more changes on top can complicate it more

Current proposal in kep.yaml is alpha in 1.25, and we can decide in the 1.25 timeframe if that still makes sense. But we should merge soon so external components know what's coming.

aojea commented:
> > +1 but I suggest to postpone one release waiting for the dockershim feedback, it seems is causing more drama than expected and adding more changes on top can complicate it more
> 
> Current proposal in kep.yaml is alpha in 1.25, and we can decide in the 1.25 timeframe if that still makes sense. But we should merge soon so external components know what's coming.

I wanted to mean 1.25, I missed that detail
lgm

danwinship commented:
Updated:
- squashed, removed UNRESOLVED section
- reverted back to original LoadBalancerServices proposal, which fixes the iptables-ordering problem and resolves Tim's objection to splitting up the service logic between nat and filter.
- explained the martian packet problem and clarified that we should probably have kubelet keep creating the rule protecting against that, even though it's not kubelet's problem :slightly_frowning_face: 

> I wanted to mean 1.25, I missed that detail

yeah, I didn't want to try to stack another iptables proxier change on top of what's already in 1.24...

danwinship commented:
@thockin can we merge this and then merge https://github.com/kubernetes/kubernetes/pull/109059 to add the `KUBE-IPTABLES-HINT` chain for 1.24?

You had previously "mostly LGTM"ed it, and I added some "UNRESOLVED" sections making it explicit that the low-level details are still subject to change. But I'd like to get this as the plan of record so that we can (1) add `KUBE-IPTABLES-HINT` in 1.24, and (2) do a blog post about the upcoming iptables rules reorg and what third-party components need to know (which I've already started writing).

squeed commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,802 @@
+kube-proxy. But because they are generally doing much less IPTables
+processing than kube-proxy does, they could fairly easily be rewritten
+to not use the packet mark, and just have slightly-redundant
+`PREROUTING` and `POSTROUTING` rules instead.
```
Yeah, we can fix CNI to not use a mark, though it will have a lot of duplicated logic.

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
````diff
@@ -0,0 +1,839 @@
+```
+<<[UNRESOLVED chain-renaming ]>>
+
+- (Maybe) Rename kube-proxy's `KUBE-MARK-MASQ` and `KUBE-POSTROUTING`
````
At this point it seems pretty hostile.  Hyrum's law says this is part of our API now. :(

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,839 @@
+it, then we need to keep their code in sync, which is one of the
+things we're supposed to not have to do any more with this KEP.
+
+We should consider deprecating this rule in kubelet. Perhaps for a
```
+1

thockin commented on keps/sig-network/3178-iptables-cleanup/README.md:
````diff
@@ -0,0 +1,839 @@
+```
+<<[UNRESOLVED cleanup ]>>
+
+It's possible we don't actually want to clean up the old chains; we
````
I agree with this, I think

thockin reviewed (commented):
Thanks!

/lgtm
/approve

danwinship commented:
/assign @johnbelamaric 
for PRR approval
(and can you `/lgtm` too because I just had to re-push with the prod-readiness file filled in, but it's otherwise unchanged from the version Tim lgtm'ed)

aojea commented:
/lgtm

johnbelamaric commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,764 @@
+
+###### What specific metrics should inform a rollback?
+
+Any failures would be the result of third-party components being
```
I understand that we can't predict what some other tool will do. But perhaps we could give some hints as to where folks could look. For example, what components (CNI, netpol implementation, etc.) are likely to have potential errors or failures?

johnbelamaric commented on keps/sig-network/3178-iptables-cleanup/README.md:
```diff
@@ -0,0 +1,764 @@
+There is no simple way to do this because if the feature is working
+correctly there will be no difference in externally-visible behavior.
+(The generated iptables rules will be different, but the _effect_ of
+the generated iptables rules will be the same.)
```
This feature isn't really "used by workloads" so much as it is "configured by the administrator", so you can say something that effect.

johnbelamaric reviewed (approved):
/approve

It's OK as is but a few more hints for users to know if this is causing issues would be nice.

johnbelamaric commented:
by the way, I don't see this on the enhancements spreadsheet for 1.25, please add it there.

danwinship commented:
/retest
